libvirtd plugin with DO go package

dahendel opened the pull request:
This libvirt plugin uses digital oceans libvirt-go package to pull metrics. As of now it pulls VM metrics and disk metrics. Network metrics are being worked on.

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [ ] Has appropriate unit tests.

danielnelson commented on plugins/inputs/libvirtd/libvirtd.go:
```diff
@@ -0,0 +1,194 @@
+)
+
+// MockPlugin struct should be named the same as the Plugin
+type LibVirtD struct {
```
I would just call this `Libvirt`

danielnelson commented on plugins/inputs/libvirtd/libvirtd.go:
```diff
@@ -0,0 +1,194 @@
+		}
+	}
+
+	l.DiskDevices = diskDevices
```
Return values instead of stashing them on the struct

danielnelson commented on plugins/inputs/libvirtd/libvirtd.go:
```diff
@@ -0,0 +1,194 @@
+	tags["name"] = dom.Name
+	tags["id"] = string(dom.ID)
+
+	acc.AddFields("libvirtd_vmdata", fields, tags)
```
I would call this `libvirt_domain`, we want to stick to the upstream terminology if possible.

danielnelson commented on plugins/inputs/libvirtd/README.md:
```diff
@@ -0,0 +1,91 @@
+1513349120000000000
+
+
+>libvirtd_vmdata,name=r-2828-QA,id=̢,host=njcloudhost.dev.ena.net branch_misses="&{6 0}",ref_cpu_cycles="&{6 0}",
```
Obviously some type conversions still to do...

danielnelson commented on plugins/inputs/libvirtd/libvirtd.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// Find all the disks in the XML data
+func (l *LibVirtD) getDiskDevices(dom libvirt.Domain, acc telegraf.Accumulator) {
```
I wonder if instead of reading through the domain definition to pull out the disks we should just iterate through the  volumes.  This is more familiar to me as a libvirt user, since volumes are one of main parts of the virsh command, while disks are somewhat hidden inside the domain.  Maybe we can avoid parsing xml as a bonus.

dahendel replied:
In our use case for example, we use ceph and none of the vol commands list our disks. I could not find another way around getting the data without looping through the XML. I am rather new to libvirt so there may be another way that I have not figured out yet. Side note, it seems like most virsh commands are built into the libvirtd-go lib.

tubemeister commented:
Hi. I've been playing with your plugin and it works, but it also seems to leak filehandles (the libvirt sock) at an alarming rate, growing by 6-700 open files per minute on my test server and crashing every few hours due to running out.

I've added a c.Close() before the final return nil, this seems to work. I still get metrics and the open file count stays at a reasonable level.

This was my first ever line of Go code, so by all means sanity check the hell out of it. ;-) 

Thanks!

tubemeister commented:
Oh, and you seem to have a typo in the volumes metric: livbirt_dom_volumes. :-)

dahendel commented:
@tubemeister thanks! I will get that changed.

dahendel commented:
I updated the Gather func with defer c.Close() and corrected the typo. However I am having trouble getting it to build because of the Microsoft lib.
```
# github.com/Microsoft/ApplicationInsights-Go/appinsights
../../Microsoft/ApplicationInsights-Go/appinsights/telemetry.go:382:27: multiple-value uuid.NewV4() in single-value context
../../Microsoft/ApplicationInsights-Go/appinsights/telemetry.go:412:23: multiple-value uuid.NewV4() in single-value context
../../Microsoft/ApplicationInsights-Go/appinsights/telemetrycontext.go:88:52: multiple-value uuid.NewV4() in single-value context
```

danielnelson commented:
I think it will be easiest to update the Godeps file by hand right now.  I merged the separate Windows and Linux Godeps into one but gdm doesn't like it.  I would reset it to the latest version on master and then add your new deps in by hand.

dahendel commented:
Closing in preference to #2560
